scripts/remove-unregistered-users.ts

```diff
@@ -0,0 +1,125 @@
+  usageParams: `CREDENTIAL_PATH`,
+  exampleParams: `fooAccountKey.json`,
+});
+
+const [credentialPath] = process.argv.slice(2);
+if (!credentialPath) {
+  usage();
+}
+
+if (!checkFileExists(credentialPath)) {
+  console.error("Credential file path does not exists:", credentialPath);
+  process.exit(1);
+}
+
+const { project_id: projectId } = parseCredentialFile(credentialPath);
+
+if (!projectId) {
+  console.error("Credential file has no project_id:", credentialPath);
+  process.exit(1);
+}
+
+initFirebaseAdminApp(projectId, {
+  credentialPath: credentialPath
+    ? resolve(__dirname, credentialPath)
+    : undefined,
+});
+
+// 1000 is the maximum number of users firestore can fetch. check .listUsers()'s documentation for more information.
+// It is also 1000 by default but I decided to set it explicitly to avoid confusion.
+const MAX_USERS = 1000;
+
+const getAuthUsers = async (maxUsers: number, pageToken?: string) =>
+  await admin.auth().listUsers(maxUsers, pageToken);
+
+(async () => {
+  const totalAuthUsers: admin.auth.UserRecord[] = [];
+
+  console.log("Fetching firestore auth users...");
+  const { users: authUsers, pageToken } = await getAuthUsers(MAX_USERS);
+  totalAuthUsers.push(...authUsers);
+
+  // @debt This is going to make another request with offset if there are the maximum users returned by the previous request. This should be using recursion.
+  if (authUsers.length >= MAX_USERS) {
+    console.log(
+      `Auth users are more than ${MAX_USERS}. Fetching additional batch of firestore auth users...`
+    );
+    const { users: additionalAuthUsers } = await getAuthUsers(
+      MAX_USERS,
+      pageToken
+    );
+    totalAuthUsers.push(...additionalAuthUsers);
+  }
+
+  const userDocIds: string[] = await admin
+    .firestore()
+    .collection("users")
+    .get()
+    .then((result) => {
+      console.log("Firestore users successfully fetched.");
+      return result.docs.map((doc) => doc.id);
+    })
+    .catch((err) => {
+      console.error(
+        "Failed when trying to fetch the users collection, skipping...",
+        err
+      );
+      return [];
+    });
+
+  const unregisteredUsers: string[] = [];
+
+  totalAuthUsers.forEach((user) => {
+    if (!userDocIds.includes(user.uid)) {
+      unregisteredUsers.push(user.uid);
+    }
+  });
+
+  const numberOfUnregisteredUsers = unregisteredUsers.length;
+
+  if (!numberOfUnregisteredUsers) {
+    console.error(
+      "There are no unregistered users in the database, end of script."
+    );
+    return;
+  }
+
+  console.error(`Deleting ${numberOfUnregisteredUsers} users...`);
+
+  await admin
+    .auth()
+    .deleteUsers(unregisteredUsers)
+    .then(() => {
+      console.error(
+        `${numberOfUnregisteredUsers} users have been successfuly deleted.`
+      );
+    })
+    .catch((err) => {
+      console.error(
+        "Failed when trying to delete the unregistered users, end of script.",
+        err
+      );
+    });
```

Comment:
We're already using `await`, instead of doing these in `.then` we should just do it on the next line after the `await`.

The `.catch` would only really make sense to re-throw the error with a better message so that the final 'catastrophic failure' `.catch` handles it and logs it.

Reply:
IMO it doesn't really matter because this is a script. And the rest of the scripts are following this same pattern. If you really want this change I will do it but I don't think it's something huge.

Reply:
I’m not going to block this PR over it, but if existing scripts do it it’s only for legacy reasons, none of the newer scripts I’ve written/seen do it this way, it makes the code simpler overall, and it makes it easier to write in the first place; so using our best practice patterns means less bad examples for people to copy from and thibk are the way to do things in future, so I’d prefer it.